incubator/hnc/pkg/stats/metrics.go

```diff
@@ -27,18 +49,90 @@ var (
 		Description: "The number of HierConfig writes happened during HierConfig reconciliations",
 		Aggregation: ocview.LastValue(),
 	}
+
+	namespaceWritesView = &ocview.View{
+		Name:        "hnc/reconcilers/hierconfig/namespace_writes_total",
+		Measure:     namespaceWritesTotal,
+		Description: "The number of namespace writes happened during HierConfig reconciliations",
+		Aggregation: ocview.LastValue(),
+	}
+
+	objectReconcileTotalView = &ocview.View{
+		Name:        "hnc/reconcilers/object/total",
+		Measure:     objectReconcileTotal,
+		Description: "The total number of object reconciliations happened",
+		Aggregation: ocview.LastValue(),
+		TagKeys:     []tag.Key{KeyGroupKind},
+	}
+
+	objectReconcileConcurrentView = &ocview.View{
+		Name:        "hnc/reconcilers/object/concurrent_peak",
+		Measure:     objectReconcileConcurrent,
+		Description: "The peak concurrent object reconciliations happened in the last reporting period",
+		Aggregation: ocview.LastValue(),
+		TagKeys:     []tag.Key{KeyGroupKind},
+	}
+
+	objectWritesView = &ocview.View{
+		Name:        "hnc/reconcilers/object/object_writes_total",
+		Measure:     objectWritesTotal,
+		Description: "The number of object writes happened during object reconciliations",
+		Aggregation: ocview.LastValue(),
+		TagKeys:     []tag.Key{KeyGroupKind},
+	}
+)
+
+// Declare periodic peak variables for concurrent reconciliations.
+var (
+	peakConcurrentHierConfigReconcile counter
+	peakConcurrentObjectReconcile     map[schema.GroupKind]counter = make(map[schema.GroupKind]counter)
 )
 
 func startRecordingMetrics() {
 	log := ctrl.Log.WithName("metricsRecorder")
 
-	// Register the view. It is imperative that this step exists,
+	// Register the views. It is imperative that this step exists,
 	// otherwise recorded metrics will be dropped and never exported.
-	if err := ocview.Register(hierReconcileView, hierWritesView); err != nil {
+	if err := ocview.Register(
+		hierReconcileTotalView,
+		hierReconcileConcurrentView,
+		hierWritesView,
+		namespaceWritesView,
+		objectReconcileTotalView,
+		objectReconcileConcurrentView,
+		objectWritesView,
+	); err != nil {
 		log.Error(err, "Failed to register the views")
 	}
+
+	// Start a new thread to reset periodic peak.
+	go resetPeriodicPeak()
 }
 
-func recordMetricsInt64(c counter, ms *ocstats.Int64Measure) {
-	ocstats.Record(context.Background(), ms.M(int64(c)))
+// recordMetric records a measurement to a predefined measure without any specific tags.
+func recordMetric(m counter, ms *ocstats.Int64Measure) {
+	ocstats.Record(context.Background(), ms.M(int64(m)))
+}
+
+// recordTagMetric inserts a tag to the context before recording a metric.
+// The tag is used to group and filter collected metrics later on.
+func recordTagMetric(m counter, ms *ocstats.Int64Measure, k tag.Key, v string) {
+	ctx, _ := tag.New(context.Background(), tag.Insert(k, v))
+	ocstats.Record(ctx, ms.M(int64(m)))
+}
+
+func resetPeriodicPeak() {
```

Comment:
I think this might make more sense as "recordPeakConcurrentReconciles" or something like that. This function doesn't _just_ reset - it actually records as well.